heartbeat/monitors/wrappers/monitorstate/tracker.go

```diff
@@ -104,3 +108,45 @@ func (t *Tracker) getCurrentState(sf stdfields.StdMonitorFields) (state *State)
 func NilStateLoader(_ stdfields.StdMonitorFields) (*State, error) {
 	return nil, nil
 }
+
+func AtomicStateLoader(inner StateLoader) (sl StateLoader, replace func(StateLoader)) {
+	mtx := &sync.Mutex{}
+	return func(currentSL stdfields.StdMonitorFields) (*State, error) {
+			mtx.Lock()
+			defer mtx.Unlock()
+
+			return inner(currentSL)
+		}, func(sl StateLoader) {
+			mtx.Lock()
+			defer mtx.Unlock()
+			inner = sl
+			logp.L().Info("Updated atomic state loader")
+		}
+}
+
+func DeferredStateLoader(inner StateLoader, timeout time.Duration) (sl StateLoader, replace func(StateLoader)) {
+	stateLoader, replaceStateLoader := AtomicStateLoader(inner)
+
+	wg := sync.WaitGroup{}
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	wg.Add(1)
+	go func() {
+		<-ctx.Done()
+
+		if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+			logp.L().Warn("Timeout trying to defer state loader")
+		}
+
+		defer cancel()
```

Comment:
why defer these both if no statements follow them? Were they intended to go at the top of the function?

Reply:
Probably a mix of an idiomatic declaration of intent and a sequential mindset when writing it 😄 I can move them to the top